### PR TITLE
[featherstone] Publish JointFeedback forces.

### PR DESCRIPTION
**Pull request by Fixit-Davide**
# 🎉 New feature
Related to the PR: https://github.com/gazebosim/gz-sim/pull/2369

Partially solves: https://github.com/gazebosim/gz-sim/issues/883

## Summary
Allow the `GetJointForce()` method in bullet-featherston to return the joint feedback Force and not 0. This allow the `joint_state_publisher` to publish data containing the force field.

## Test it
- Launch `gz sim --physics-engine gz-physics-bullet-featherstone-plugin lift_drag.sdf` simulation.
- Send a command to the model (`gz topic -t "/model/lift_drag_demo_model/joint/rod_1_joint/cmd_force" -m gz.msgs.Double  -p "data: 0.8"`).
- Echo the topic: `gz topic -e -t /world/lift_drag/model/lift_drag_demo_model/joint_state`
## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

### CODE REVIEW

File: bullet-featherstone/src/JointFeatures.cc [hunk 1]
```diff
@@ -80,16 +80,42 @@ double JointFeatures::GetJointAcceleration(
...
+      math::Vector3 axis_converted(axis[0], axis[1], axis[2]);
+      btVector3 angular = feedback->m_reactionForces.getAngular();
+      math::Vector3<double> angularTorque(
+        angular.getX(),
+        angular.getY(),
+        angular.getZ());
+      // BUG: The total force is 2 times the cmd one see:
+      // https://github.com/bulletphysics/bullet3/discussions/3713
+      results += axis_converted.Dot(angularTorque);
+      return results / 2.0;
```
Comment (iche033) [inline bullet-featherstone/src/JointFeatures.cc:105]: Is the `results / 2.0` done to workaround the bug? Note that it was recently fixed upstream: https://github.com/bulletphysics/bullet3/pull/4583. I think we also found that total force is not always 2 times than it should be (based on example in https://github.com/gazebosim/gz-physics/issues/565 and comment: https://github.com/bulletphysics/bullet3/discussions/3713#discussioncomment-657881). 

Anyways, best to guard this with a bullet version check e.g. for users using latest bullet version built from source.

```
#if BT_BULLET_VERSION < 326
  // not sure if this is always true
  return results / 2.0; 
#else
  return results;
#endif
```

File: bullet-featherstone/src/JointFeatures.cc [hunk 2]
```diff
@@ -80,16 +80,42 @@ double JointFeatures::GetJointAcceleration(
 double JointFeatures::GetJointForce(
     const Identity &_id, const std::size_t _dof) const
 {
+  double results = gz::math::NAN_D;
   const auto *joint = this->ReferenceInterface<JointInfo>(_id);
   const auto *identifier = std::get_if<InternalJoint>(&joint->identifier);
-  if (identifier)
-  {
+
+  if (identifier) {
```
Comment (iche033) [inline bullet-featherstone/src/JointFeatures.cc:87]: nit: gazebo coding style:


```suggestion
  if (identifier) 
  {
```

File: bullet-featherstone/src/JointFeatures.cc [hunk 3]
```diff
@@ -80,16 +80,42 @@ double JointFeatures::GetJointAcceleration(
...
     const auto *model = this->ReferenceInterface<ModelInfo>(joint->model);
-    return model->body->getJointTorqueMultiDof(
-          identifier->indexInBtModel)[_dof];
+    auto feedback = model->body->getLink(identifier->indexInBtModel).m_jointFeedback;
+    const auto &link = model->body->getLink(identifier->indexInBtModel);
+    results = 0.0;
+    if (link.m_jointType == btMultibodyLink::eRevolute) {
```
Comment (iche033) [inline bullet-featherstone/src/JointFeatures.cc:92]:
```suggestion
    if (link.m_jointType == btMultibodyLink::eRevolute) 
    {
```

File: bullet-featherstone/src/JointFeatures.cc [hunk 4]
```diff
@@ -80,16 +80,42 @@ double JointFeatures::GetJointAcceleration(
...
+      btVector3 angular = feedback->m_reactionForces.getAngular();
+      math::Vector3<double> angularTorque(
+        angular.getX(),
+        angular.getY(),
+        angular.getZ());
+      // BUG: The total force is 2 times the cmd one see:
+      // https://github.com/bulletphysics/bullet3/discussions/3713
+      results += axis_converted.Dot(angularTorque);
+      return results / 2.0;
+    } else if (link.m_jointType == btMultibodyLink::ePrismatic) {
```
Comment (iche033) [inline bullet-featherstone/src/JointFeatures.cc:106]:
```suggestion
    } 
    else if (link.m_jointType == btMultibodyLink::ePrismatic) 
    {
```

File: bullet-featherstone/src/JointFeatures.cc (hunk 1 above)
Comment (Fixit-Davide) [reply bullet-featherstone/src/JointFeatures.cc:105]: Yes, the `results / 2.0` part is used as an ugly workaround in order to retrieve a data that is reasonable. I didn't know that the PR was merged when I have write the code.
Probably another workaround (for bullet version < 3.26) could be `measured_torque = measured_torque - applied_torque` ? But I don't know how to retrieve the information about the applied torque from the  `jointInfo` class.

Comment (iche033) [reply bullet-featherstone/src/JointFeatures.cc]: I just did some quick tests and saw that bullet clears the applied joint forces and torques after each step so I think we're not able to retrieve the applied torque at this point (e.g. `model->body->getJointTorqueMultiDof` just returns 0). We could cache the applied torque in `JointInfo` struct but we would need to clear them before every step. For now, I think we can just add a comment about this bug affecting bullet versions < 326 - basically what you had before, a link to https://github.com/bulletphysics/bullet3/discussions/3713 or https://github.com/gazebosim/gz-physics/issues/565.

**Comment by iche033** [comment]
CI picked up a few cpplint warnings:

```
  /github/workspace/bullet-featherstone/src/JointFeatures.cc:87:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  /github/workspace/bullet-featherstone/src/JointFeatures.cc:90:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/bullet-featherstone/src/JointFeatures.cc:95:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

**Review by iche033 — APPROVED** [review]

**Comment by Fixit-Davide** [comment]
Any changes needed in order to merge? I can write the test for https://github.com/gazebosim/gz-sim/pull/2369 afterward.

**Comment by iche033** [comment]
I'll go ahead and merge this first

**Pull request merged**